src/index.ts

```diff
@@ -31,104 +36,178 @@ const rgbCallExpRegex =
 const hslCallExpRegex =
   /hsl\(\s*(\d{1,3})\s*,\s*(\d{1,3})%\s*,\s*(\d{1,3})%\s*(,\s*0?\.\d+)?\)/;
 
-function colorPickersDecorations(view: EditorView) {
-  const widgets: Array<Range<Decoration>> = [];
+function enter(
+  syntaxTree: Tree,
+  from: number,
+  to: number,
+  typeName: string,
+  doc: Text,
+): WidgetOptions | Array<WidgetOptions> | null {
+  switch (typeName) {
+    case 'AttributeValue': {
+      const innerTree = syntaxTree.resolveInner(from, 0).tree;
+
+      if (!innerTree) {
+        return null;
+      }
 
-  for (const range of view.visibleRanges) {
-    syntaxTree(view.state).iterate({
-      from: range.from,
-      to: range.to,
-      enter: ({type, from, to}) => {
-        if (type.name === 'CallExpression') {
-          const callExp: string = view.state.doc.sliceString(from, to);
-          if (callExp.startsWith('rgb')) {
-            const match = rgbCallExpRegex.exec(callExp);
-
-            if (!match) {
-              return;
-            }
+      const overlayTree = innerTree.prop(NodeProp.mounted)?.tree;
 
-            const [_, r, g, b, a] = match;
-            const color = rgbToHex(r, g, b);
-
-            const widget = Decoration.widget({
-              widget: new ColorPickerWidget({
-                colorType: ColorType.rgb,
-                color,
-                from,
-                to,
-                alpha: a || '',
-              }),
-              side: 1,
-            });
+      if (overlayTree?.type.name !== 'Styles') {
+        return null;
+      }
+
+      const ret: Array<WidgetOptions> = [];
+      overlayTree.iterate({
+        from: 0,
+        to: overlayTree.length,
+        enter: ({ type, from: overlayFrom, to: overlayTo }) => {
+          const maybeWidgetOptions = enter(
+            syntaxTree,
+            // We add one because the tree doesn't include the
+            // quotation mark from the style tag
+            from + 1 + overlayFrom,
+            from + 1 + overlayTo,
+            type.name,
+            doc,
+          );
 
-            widgets.push(widget.range(from));
+          if (maybeWidgetOptions) {
+            if (Array.isArray(maybeWidgetOptions)) {
+              throw new Error('Unexpected nested overlays');
+            }
 
-            return;
+            ret.push(maybeWidgetOptions);
           }
-          if (callExp.startsWith('hsl')) {
-            const match = hslCallExpRegex.exec(callExp);
+        },
+      });
 
-            if (!match) {
-              return;
-            }
+      return ret;
+    }
```

Comment:
this is the main change, everything else is a small refactor to allow for recursive calls.